### PR TITLE
[DMR-9] Replacing cookcc based parser with new one to support JDK8 and above

### PR DESCRIPTION
**Pull request by ropalka**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -40,6 +40,7 @@
     <properties>
         <maven.compiler.target>1.6</maven.compiler.target>
         <maven.compiler.source>1.6</maven.compiler.source>
```
Comment (ctomc) [inline pom.xml]: 1.8 probably :)

File: src/main/java/org/jboss/dmr/stream/ModelReader.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+     */
+    boolean isNumber();
+
+    /**
+     * Converts available context data to <code>Number</code> instance.
+     * Users have to call {@link #next()} and should call {@link #isNumber()} before calling this method.
+     * @return number value the parsing cursor is pointing to
+     * @exception NumberFormatException if DMR number is not convertible to <code>number</code>
+     */
+    Number getNumber();
```
Comment (dmlloyd) [inline src/main/java/org/jboss/dmr/stream/ModelReader.java]: I don't like generalizing this.  It's better to have separate getInt(), getLong() which do not box and getBigInteger() which returns BigInteger.

File: src/main/java/org/jboss/dmr/stream/ModelReaderImpl.java [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+    private char[] buffer = new char[ 1024 ];
+    private int position;
+    private int limit;
+    private int numberOffset;
+    private int numberLength;
+    private int stringOffset;
+    private int stringLength;
+    private byte[] bytesValue;
+    private ModelType typeValue;
+    private Number numberValue;
```
Comment (dmlloyd) [inline src/main/java/org/jboss/dmr/stream/ModelReaderImpl.java]: Using separate fields for the int or long value may reduce allocation churn a bit while reading the model, by not maintaining boxed Integer and Long instances.

File: src/main/java/org/jboss/dmr/stream/ModelType.java [hunk 4]
```diff
@@ -20,11 +20,25 @@
...
-
-
-public class JSONParser extends CookCCByte {
+package org.jboss.dmr.stream;
 
+/**
+ * DMR types.
+ * @author <a href="mailto:ropalka@redhat.com">Richard Opalka</a>
+ */
+public enum ModelType {
```
Comment (dmlloyd) [inline src/main/java/org/jboss/dmr/stream/ModelType.java]: This still needs to be unified with the "real" ModelType.

**Comment by dmlloyd** [comment]
Other than the above, it looks good!

**Comment by dmlloyd** [comment]
The cure might be worse than the disease.  Is there any way we can just have stubs here?

@bstansberry WDYT?

**Comment by ropalka** [comment]
You mean unfunctional stubs?

**Comment by ropalka** [comment]
I had a quick look. Unfortunately most methods from generated parsers are public or protected and generated classes are public. This makes it hard to implement the proxy behaviour. IMHO it's much more simpler (minimum effort is required) to include old classes + old generated classes and just deprecate them. WDYT?

**Comment by dmlloyd** [comment]
If it makes it impossible to build on Java 8, that's a strong point against the idea.

**Comment by ropalka** [comment]
It builds on JDK8 ATM because I included old generated parser sources in the src/main/folder.

**Comment by ctomc** [comment]
I think we should just keep the stubs of public methods.

**Comment by ropalka** [comment]
To be precise all methods of generated Parser and JSONParser are public :-(

**Comment by ctomc** [comment]
Not all, there are many protected ones :)

**Comment by bstansberry** [comment]
The last commit seems ok to me, but I think it's fine to remove the cookcc annotations and hence the pom dependency.

I don't see a point to unfunctional stubs, as I can't think why someone would compile against these but not use them at runtime. Eliminating protected methods doesn't buy much IMO. We should add an @deprecated javadoc saying these were never meant for external use, may be removed at any time, and will not be maintained. So no bug fixing. We're just being dutiful about deprecating before removing.

**Comment by bstansberry** [comment]
Side note one this: there have been 6 changes to JsonParserImpl since its initial commit, and 2 of those were a change and then a revert of the change. The cookcc dep version has never been changed since first added, so presumably there's been little change in the generated classes either. So I think the likelihood of feeling compelled to do any maintenance on these classes before we remove them is almost nil. That's the only reason I suggested being courteous (aka anal) and deprecating them for a bit.

**Comment by puntogil** [comment]
Is possible push this branch/changes  as/in a  stable release?

**Comment by bstansberry** [comment]
@puntogil it will be a while before it is in a .Final release, as it should bake in WildFly/WildFly Core's dev branches for a bit. I'll do a beta soon though.

**Comment by bstansberry** [comment]
@ropalka the docs stuff is awesome! So I feel guilty, but can you rename dmr-protocol and json-protocol subdirs to dmr-format / json-format?

I'm sure I'll link to this excellent stuff in the WildFly docs and I don't want to imply that DMR is a communications protocol. A lot of people working around WildFly already misuse the term 'DMR' when they mean WildFly management in general, or JBoss Remoting based WildFly management or....

**Comment by ropalka** [comment]
Here you go @bstansberry

**Comment by bstansberry** [comment]
As I commented on DMR-9:

Building core with a jboss-dmr snapshot with this in it fails in the cli module testsuite.

It looks like the output from ModelNode.fromString("a64180c0)e7187af86b435229de904104"") differs from previous versions. Note the trailing '"' in the input.

**Comment by ropalka** [comment]
The test is buggy @bstansberry, see: https://github.com/ropalka/wildfly-core/commit/aa63aa938762cb0703cc435bf57b81792f8dfb05

**Pull request merged**